Implement Promise.prototype.finally Fixes #3520

rhuanjl opened the pull request:
This PR implements Promise.prototype.finally.

Fixes #3520
Relevant ECMASpec: https://tc39.github.io/ecma262/#sec-promise.prototype.finally

**Notes:**
1. I worked out how to do this by reading the code for Promise.prototype.catch and the internal AsyncSpawnExecutorFunction - there is a chance that some of what I've done isn't quite right though it does all seem to work.
2. I've run this against the relevant testcases from test262 and it passes them.
3. It probably could do with a few more CI tests - I added what seemed good to me based on what was there for other promise methods.
4. The large size of the diff is as I had to regenerate the bytecode for built in methods after adding the entry point for finally to JnDirectFields.h

**cc:** @dilijev @ljharb  

Thanks to my friend @fatcerberus for running the RegenerateBytecode script for me.

ljharb commented on test/es6/ES6Promise.js:
```diff
@@ -82,6 +82,13 @@ var tests = [
+            assert.isFalse(descriptor.enumerable, "Promise.prototype.finally.enumerable === false");
+            assert.isTrue(descriptor.configurable, "Promise.prototype.finally.configurable === true");
+            assert.areEqual('function', typeof descriptor.value, "typeof Promise.prototype.finally === 'function'");
+            assert.areEqual(1, Promise.prototype.finally.length, "Promise.prototype.finally.length === 1");
```
can you also add an assert that the `.name` is `"finally"`?

rhuanjl replied:
Done - and added equivalents for .then and .catch as they didn't exist either.

MSLaguana commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,154 @@ namespace Js
+        
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        Assert(JavascriptPromiseThenFinallyFunction::Is(function));
```
Actually, since you do that in the `JavascriptPromiseThenFinallyFunction::FromVar` method, you should just call that I think.

fatcerberus replied:
Out of curiosity (and for future reference on my part 😄), what is the criteria for whether something should be a regular assert vs. a FailFast assert?  CC codebase has both kinds.

MSLaguana replied:
If it's something that could lead to type confusion / executing wrong code, it should typically be a failfast to avoid security issues. If it is something that shouldn't happen, and we'd like to know in tests when it happens so we can fix/understand it, but is recoverable then a plain assert is fine.

rhuanjl replied:
My thinking was that there was no existing code path to get there without that Assert being true and so no need to add any extra cost at runtime and future regressions could be detected by the Assert before making it into any releases. I've made the change though.

MSLaguana commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,154 @@ namespace Js
+
+
+        //8. Return ? Invoke(promise, "then", <<valueThink>>)
+        JavascriptPromise* promise = UnsafeFromVar(promiseVar);
```
Why do you convert from `Var` to `JavascriptPromise` here? It looks to me like you just access properties on the object, none of the native methods of `JavascriptPromise`.

rhuanjl replied:
This was mistake - I rewrote the logic in this function several times and this was a holdover from an earlier version, fixed.

fatcerberus commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,154 @@ namespace Js
+        JavascriptPromiseThunkFinallyFunction* Thunk = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, args[1], This->GetShouldThrow());
+
+
+        //8. Return ? Invoke(promise, "then", <<valueThink>>)
```
Typo in comment: Should be `valueThunk`, not "think"

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+
+        //5. If IsCallable(onFinally) is false
+        //a. Let thenFinally be onFinally
+        //b. let catchFinally be onFinally 
```
nitpick: 'let' here is not capitalized and the a and b items are not indented.

rhuanjl replied:
fixed

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+        //  d. Set thenFinally and catchFinally's [[OnFinally]] internal slots to onFinally.
+
+        Var thenFinally;
+        Var catchFinally;
```
I'd like if these were initialized to `nullptr` and have an assert after the conditional that each is not `nullptr` anymore. Should cause no overhead in release build.

rhuanjl replied:
fixed

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+        
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+
+        JavascriptPromiseThenFinallyFunction* This = JavascriptPromiseThenFinallyFunction::FromVar(function);
```
I'm not a big fan of naming this `JavascriptPromiseThenFinallyFunction*` as `This`. Can we call it `finallyFunction` or `thenFinallyFunction` or something?

rhuanjl replied:
OK (sorry Javascript is my native language was trying to make it look like this.property)

boingoing replied:
No worries, we just don't usually use this naming convention in our codebase. 👍

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+        Var result = CALL_FUNCTION(scriptContext->GetThreadContext(), This->GetOnFinally(), CallInfo(CallFlags_Value, 1), library->GetUndefined());
+
+        //4. Let C be F.[[Constructor]]
+        //5.Assert IsConstructor(C)
```
nitpick: A space here

rhuanjl replied:
fixed

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+
+        //7. Let valueThunk be equivalent to a function that returns value
+        //OR 7. Let thrower be equivalent to a function that throws reason
+        JavascriptPromiseThunkFinallyFunction* Thunk = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, args[1], This->GetShouldThrow());
```
Same naming nitpick as above. Can we not use `Thunk` here but something like `thunkFinallyFunction`?

rhuanjl replied:
changed

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+        JavascriptPromiseThunkFinallyFunction* Thunk = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, args[1], This->GetShouldThrow());
+
+        //8. Return ? Invoke(promise, "then", <<valueThunk>>)
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(promiseVar);
```
Are we sure `UnsafeFromVar` is ok here? Maybe regular `FromVar` is better unless `CreateResolvedPromise` has a guarantee.

rhuanjl replied:
I think this is fine - CreateResolvedPromise uses a FromVar on the constructor to check it's safe then uses it to generate the promise. For this promise to be invalid therefore without hitting the FailFast in CreateResolvedPromise would require CreateResolvedPromise to be broken. - Not changed - will change if you still want me to though.

boingoing replied:
The `FromVar` in `CreateResolvedPromise` is done to the `constructor` object, though. Here we are converting the promise object which was returned from `CreateResolvedPromise` which could be any `Var`, couldn't it? It's just whatever was returned from the `constructor` when we called it in `CreatePromiseCapabilityRecord`. Since we're calling as-constructor, the return should be an object, I guess. I do wonder if it would be possible to get it to be a non-object. Seems might be possible. I think we should be cautious and follow the `Invoke` rabbit-hole down to 7.3.2 GetV(V,P) ([https://tc39.github.io/ecma262/#sec-getv](url)) where we should do `ToObject` on `promise`. We do something very similar in `EntryAll` on a newly constructed promise Var.

rhuanjl replied:
OK I've put in an AssertOrFailFast for it being an object, I think that should do the job? If not, not quite sure what you wanted.

fatcerberus replied:
@rhuanjl I think he just wanted you to change the UnsafeFromVar to a FromVar.

boingoing replied:
Sorry for the confusion - I said a lot of words.

When I saw this code at first I wondered if UnsafeFromVar would be actually safe because promiseVar is a Var and it isn't clear that it always must be a RecyclableObject. That's why I thought FromVar would be safer because it will check at runtime.

After taking a look further, I believe it may be possible for CreateResolvedPromise to return a non-object because the constructor may be overridden via Symbol.species. If that's the case, we should actually throw a TypeError while doing Invoke according to ECMA. Invoke calls GetV which performs ToObject on V. Does that make sense? 🤔

I believe the conversion should look like this:
```
RecyclableObject* promise = JavascriptOperators::ToObject(promiseVar, scriptContext);
```

rhuanjl replied:
Yep that makes sense - well that aspect of the spec does not seem sensible at all but ah well. Will re-push after re-building and re-running tests.

boingoing replied:
Many aspects of the spec do not seem sensible. Thanks for your efforts. 😄

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+
+        //7. Let valueThunk be equivalent to a function that returns value
+        //OR 7. Let thrower be equivalent to a function that throws reason
+        JavascriptPromiseThunkFinallyFunction* Thunk = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, args[1], This->GetShouldThrow());
```
We need to make sure `args[1]` exists. This should check `args.Info.Count` and pass `undefined` if the then finally function was called without `value`.

rhuanjl replied:
Oops, I'd only been thinking of the normal usage pattern - forgot you could call this with no argument by overwriting .then(); Fixed.

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+        RecyclableObject* undefinedVar = scriptContext->GetLibrary()->GetUndefined();
+
+        return CALL_FUNCTION(scriptContext->GetThreadContext(),
+            func, Js::CallInfo(CallFlags_Value, 3),
```
Looks like we should just be passing 2 arguments here - `this` and `valueThunk`, right? I think you can remove `undefinedVar`.

rhuanjl replied:
Oops, fixed.

boingoing commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,147 @@ namespace Js
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+
+        JavascriptPromiseThunkFinallyFunction* This = JavascriptPromiseThunkFinallyFunction::FromVar(function);
```
One more nitpick about `This`, can we name it something like `thunkFinallyFunction`?

rhuanjl replied:
changed.

boingoing commented on lib/Runtime/Library/JavascriptLibrary.cpp:
```diff
@@ -6929,6 +6933,30 @@ namespace Js
+        Assert(scriptContext->GetConfig()->IsES6PromiseEnabled());
+        
+        FunctionInfo* functionInfo = RecyclerNew(this->GetRecycler(), FunctionInfo, entryPoint);
+        DynamicType* type = CreateDeferredPrototypeFunctionType(entryPoint);
```
Can you use `GetDeferredAnonymousFunctionTypeHandler()` to avoid having to set the `name` property here?

rhuanjl replied:
How did I not notice that function... I spent 10-15 minutes searching JavascriptLibrary.cpp for a way to do that but didn't find it. Changed.

rhuanjl replied:
Note, this comment has moved down to the other CreateDeferredPrototypeFunctionType call not where you originally placed it.

rhuanjl commented:
@boingoing thanks for the review - I've made the changes and re-pushed. I've also added an extra test case to confirm that thenFinally handles the situation of being called without an argument correctly (and supplies the undefined value).

Will update and squash later on assuming you're happy with my updates.

rhuanjl commented:
@boingoing Thanks again for the review I've fixed all your points now, the commit "Responding to feedback" contains all responses to your points. Please let me know when good to squash.

boingoing commented:
@rhuanjl Looks good to me. Thanks a lot for responding to those feedback and sorry for being a bit pedantic about it.

If you squash these commits, I will begin the process of testing internally and shepherd the change through our checkin process.

rhuanjl commented:
Thanks @boingoing squashed and re-pushed. Did @dilijev wanted to review as well?

dilijev commented:
Taking a look.

kfarnung commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,159 @@ namespace Js
         return CreateThenPromise(promise, fulfillmentHandler, rejectionHandler, scriptContext);
     }
 
+    // Promise.prototype.finally as described in the draft of ES 2018 Section 25.4.5.3
```
nit:  ECMA section numbers aren't immutable, but the anchors are considered to be.  Use `#sec-promise.prototype.finally` instead of a section number.

rhuanjl replied:
I was copying what was already in the file on this one - most of the Promise methods have comments referring to section numbers per ES2015 (which are also mostly out of date now) if I'm going to fix this point assume I should fix all of them at the same time?

dilijev replied:
Might as well fix this one now and follow up with the others 

---
In reply to: [167731238](https://github.com/Microsoft/ChakraCore/pull/4668#discussion_r167731238) [](ancestors = 167731238)

boingoing replied:
If we're going to update the other section numbers, I'd like to make sure we also do an audit to make sure no behavior in the functions has changed between spec versions. Then we can be sure the implementation holds to the new spec version and update the spec reference at the same time. So, probably best to follow-up on the other methods.

dilijev commented:
Looks like we need tests of behavior (besides throwing behavior) for `.finally(<non-callable>)`

dilijev commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,159 @@ namespace Js
+        }
+
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(args[0]);
```
>UnsafeFromVar(args[0]); [](start = 54, length = 23)

Is it okay to call UnsafeFromVar(args[0]) here? -- do we know for sure this has to be a RecyclableObject at this point? It is possible to call with this as, say, a number?

dilijev replied:
Or, because the check above covers that it is an object -- is it necessarily a RecyclableObject or could it be something else (I'm not sure the answer to this question)

---
In reply to: [167733902](https://github.com/Microsoft/ChakraCore/pull/4668#discussion_r167733902) [](ancestors = 167733902)

boingoing replied:
We did a check above to make sure we have an args[0] and that it is an object. Should be guaranteed here.

rhuanjl replied:
There's a type error thrown a couple of lines above if the arg is not an Object.

I'll also add a new test case that confirms that a type error is thrown if this is called with any non-promise object.

dilijev commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,159 @@ namespace Js
+
+        Assert(thenFinally != nullptr && catchFinally != nullptr);
+
+        //7. Return ? Invoke(promise, "then", « thenFinally, catchFinally »).
```
>« thenFinally, catchFinally » [](start = 46, length = 29)

Please remove non-ascii characters from comments (this has caused problems when compiling on systems in some non-English locales in the past). You could replace these with `<<` and `>>` for example.

dilijev replied:
We should have a check for this. Added a task: [https://github.com/Microsoft/ChakraCore/issues/4672](https://github.com/Microsoft/ChakraCore/issues/4672)

---
In reply to: [167734493](https://github.com/Microsoft/ChakraCore/pull/4668#discussion_r167734493) [](ancestors = 167734493)

rhuanjl replied:
Will do, sorry about this, I'd copied and pasted these lines from the spec

dilijev replied:
No worries, common gotcha.

dilijev commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -559,6 +559,159 @@ namespace Js
+        JavascriptPromiseThenFinallyFunction* thenFinallyFunction = JavascriptPromiseThenFinallyFunction::FromVar(function);
+
+        //1. Let onFinally be F.[[OnFinally]]
+        //2. Assert: IsCallabale(onFinally)=true
```
>IsCallabale [](start = 21, length = 11)

nit: typo?

rhuanjl commented:
@dilijev I've fixed the issues in the comments and I've added some new test cases, some to cover .finally(not_callable) and one to ensure that not_promise_object.finally() throws a type error and does not crash.

dilijev commented:
@rhuanjl looks like there's a conflict with the latest version of bytecode. You'll want to merge with master and regenerate bytecode before the final squash.

fatcerberus commented:
I wonder if there should be a test for what happens if a promise has been rejected and the resulting `.finally()` continuation throws.  According to ES spec (unless I misread it) in this case the chained promise should reject with whatever you threw instead of the original rejection reason.

fatcerberus commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -560,6 +560,159 @@ namespace Js
+
+        JavascriptPromiseThunkFinallyFunction* thunkFinallyFunction = library->CreatePromiseThunkFinallyFunction(EntryThunkFinallyFunction, valueOrReason, thenFinallyFunction->GetShouldThrow());
+
+        //8. Return ? Invoke(promise, "then", <<valueThunk>>)
```
Nit: Prevailing convention seems to be to have a space between `//` and the comment text.  In any case comments are easier to read with that space IMO.

ljharb commented:
@fatcerberus your interpretation of the spec is correct.

rhuanjl commented:
@boingoing I think this is good to go now?

I've updated to latest and addressed all review points - I've also added extra test cases for the case @fatcerberus and @ljharb mentioned.

boingoing commented:
@rhuanjl Yes, looks good to me. I will take a final look over the code and pull the branch down to do internal testing before landing. Thank you for your patience.

fatcerberus commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -560,6 +560,159 @@ namespace Js
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(args[0]);
+        //3. Let C be ? SpeciesConstructor(promise, %Promise%).
+        RecyclableObject* constructor = JavascriptOperators::SpeciesConstructor(promise, scriptContext->GetLibrary()->GetPromiseConstructor(), scriptContext);
+        //4. Assert IsConstructor(C)
```
nit: space after `//`

fatcerberus commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -560,6 +560,159 @@ namespace Js
+
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(args[0]);
+        //3. Let C be ? SpeciesConstructor(promise, %Promise%).
```
nit: space after `//`

fatcerberus commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -560,6 +560,159 @@ namespace Js
+
+        // 5. If IsCallable(onFinally) is false
+        //   a. Let thenFinally be onFinally
+        //  b. Let catchFinally be onFinally 
```
nit: missing space

fatcerberus commented on lib/Runtime/Library/JavascriptPromise.cpp:
```diff
@@ -560,6 +560,159 @@ namespace Js
+
+        Var valueOrReason = nullptr;
+
+        if(args.Info.Count > 1)
```
nit: missing space after `if`

rhuanjl commented:
@boingoing I've added some spaces I was missing that @fatcerberus noticed - I'll hold on squashing until you've done your final review.

boingoing commented:
@rhuanjl I added a commit to update the bytecode cache GUID. We need this for internal processes. Otherwise we are good to go from my perspective. Please squash it and I'll run the internal testing once more. 👍

rhuanjl commented:
@boingoing ok squashed, thank you and sorry that this has taken a lot of your time.

boingoing commented:
@rhuanjl Thanks for your contribution, this is merged! No worries about my time, by the way, you've saved us a lot of time by implementing it in the first place. 👍

rhuanjl commented:
Ok, thanks for landing this @boingoing and thanks to everyone else for the reviews.

Also thanks again @fatcerberus for regenerating the bytecode for me (couldn't see a way to do that on macOS, and have never learnt how to do anything on Windows and use excel)

ljharb commented:
Yay! Any idea when to expect it in Edge?

dilijev commented:
@rhuanjl the byte code regenerate script really just calls ch with some parameters. In theory it should work on xplat although you'd have to rewrite the logic in bash or something. This is something quick we should probably add.

dilijev commented:
@ljharb hard to say exactly, but think it will be in insiders flights in a few months -- general availability whenever the following version of windows is released.

rhuanjl commented:
@dilijev I think a key hurdle would be that the current byte code regen method requires a 32bit binary and a 64bit binary - whereas at the moment ChakraCore on xplat is 64 bit only. https://github.com/Microsoft/ChakraCore/issues/4091
